shared-libs/rules-engine/src/provider-wireup.js

```diff
@@ -135,6 +139,48 @@ module.exports = {
   },
 };
 
+let refreshQueue = Promise.resolve();
+const enqueue = callback => {
+  const listeners = [];
+  const eventQueue = [];
+  const emit = evtName => {
+    if (!listeners[evtName]) {
+      return eventQueue.push(evtName);
+    }
+    listeners[evtName].forEach(callback => callback());
+  };
+
+  emit('queued');
+  refreshQueue = refreshQueue.then(() => {
+    emit('running');
+    return callback();
+  });
+
+  refreshQueue.on = (evtName, callback) => {
+    listeners[evtName] = listeners[evtName] || [];
+    listeners[evtName].push(callback);
+    eventQueue.forEach(queuedEvent => queuedEvent === evtName && callback());
```

Comment:
Is this `eventQueue` thing to get around a race condition where `emit` is called before the caller has registered the `on` listener?

It's a bit weird because it's always queued immediately so it seems pointless to emit it, except it keeps the calling code consistent.

... some time later ...

I think it's fine, it just took me a while to wrap my head around it.

Reply:
Yea, agree it's not the most obvious flow. 
I'll add some comments.